utils/docs/pads_backannotate.1.in
.TH pads_backannotate 1 "@DATE@" "gEDA Project" @VERSION@

.SH NAME
.B pads_backannotate
\- backannotates changes from a PADS engineering change order (ECO) file to schematic

.SH SYNOPSIS
.B pads_backannotate \-h\||\-\-help

.B pads_backannotate \-V\||\-\-version

.B pads_backannotate
[options]
.I ecofile.eco infile1.sch [infile2.sch ...]


.SH DESCRIPTION

The
.I pads_backannotate
program reads an engineering change order (ECO) file generated by the PADS program
and backannotates the requested changes to a gschem schematic or schematics.
Refer to the PADS manual for a complete description of the ECO file format.

.SH OPTIONS
.TP
\-h, \-\-help
prints help message and exits
.TP
\-n, \-\-nocopy
Leave the modified schematic files in new files
whose names are generated by appending
.I ".new"
to the
original file names.  The default is to overwrite the original.
.TP
\-v, \-\-verbose
operate verbosely
.TP
\-V, \-\-version
prints version information and exits

.SH EXIT STATUS
.I pads_backannotate
will return
.B 0
if all files processed successfully,
.B non-zero
if the command did not complete successfully.

.SH BUGS
Rather than parsing and processing schematic files directly, this utility should
be rewritten using the gEDA scheme API.  This would take advantage of the existing
parsers for the schematic files.

Reference designator renaming is fully supported.  Pin and gate swapping is parsed
but not automatically applied to the schematic database.  Instead, a report listing
the manual pin and gate swapping that is required.  The other features of the ECO
file are not supported yet.

.SH SEE ALSO
gnetlist(1), gschem(1)

Also refer to "Forward/Backward Annotation Between gEDA/gaf and PADS PowerPCB" which
may be found at http://wiki.geda-project.org/geda:fbabgapp
.PP
.SH COPYRIGHT
.nf
Copyright \(co 2003-@YEAR@ gEDA Contributors.  License GPLv2+: GNU GPL
version 2 or later.  Please see the `COPYING' file included with this
program for full details.
.PP
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.


NEWS
==============================================
 'gschem and Friends' Electronic Design Suite
==============================================

Copyright (C) 1998-2015 gEDA Developers

This file documents important user-visible changes in gEDA/gaf.  For
more information, please consult the `ChangeLog' file.

Notable changes in gEDA/gaf 2.1.0.0 - 04/01/15 Rampage Edition

* This release of the Rampage Edition of the gEDA-gaf suite commemorates
  the 17th Anniversary of the first public release of gEDA-gaf. The 2.09
  series releases reflected a very aggressive development of new features
  and major enhancements in many areas. Numerous enhancements focused on
  existing features and functions that did not always work as expected.

* New editing functions provide powerful methods to improve productivity.

  A new Extend Objects functions allows linear objects to be extended to
  other objects with a simple key press.

  A new Break function allows Arcs, Boxes, Circles, Linear objects, and
  Paths, both closed and open, to be broken at user specified points.

  A Powerful new Array dialog has been added to create arrays of selected
  objects. Creating Arrays allows both verb-noun or noun-verb operations.

  A new Auto-Snap Objects to Grid editing function has been added allowing
  Selections to be quickly repositioned to the current Grid snap. The Auto
  Snap features works with all objects types, including graphical and image
  objects.

  A new bimodal Offset function provides editing methods to quickly create
  new objects at programmable distances. The distance can be specified by
  entering a value in a dialog or offsetting Selection sets to the current
  pointer position using hot-keys. The offset distance is retained between
  operations.

  A new Select Last function has been added to select the last object created
  or edited.

  The visibility of netname attributes can now be toggled while inside an
  action.

  A new RC keyword; third-button-cancel, allows user to specify a preference
  for canceling operations using the third mouse button. In addition, object
  adder modes can now be cancel by double-clicking the final entry point.

  A new Zoom to Magnification feature allows user to zoom to a specified level.

* Picture replacement operations can now be undone.

* A new Export functions allow exporting of embedded objects, including image
  objects.

* Picture objects no longer need to maintain the original aspect ratio. Aspect
  ratios of image objects can be freely modified using Grips. The Ctrl key can
  be combined to maintain the aspect ratio identically to the old behavior.
  When exporting images using the new export feature, options are provided to
  export the image at the original or the current aspect ratio and size.

* Coordinates can now be displayed in the status bar using Cartesian and vector
  formats. The format can be changed at any time, include inside of an action,
  by right clicking on the coordinates widget and selecting the format from a
  pop-up menu. Vector formats revert to the previous selected format when not
  applicable to the active context.

* Pointer button assignments can now be set via a pop-up menu by clicking on
  the associated status widget on the Status bar. Assignments can still be set
  using in the Preferences Dialog or over-ridden using RC keywords, but the
  new method allows changing assignments while inside an action.

* The status text now changes to blue when inside of an editing action.

* F-Keys have been assigned to common/standard program functions.

* Menus in Gschem have been revised for better organization with enhanced
  options. The tools menu now includes access to Version tracking and
  alternative simulations back-ends. A toggle widget has been added to
  the Options menu to toggle a new drag-can-move preference variable in
  order to enable and disable dragging moving.

  A new File/Export main menu fly-out was added for options to export both
  embedded components and pictures. Exporting pictures utilized a new custom
  Image Chooser Dialog allowing user to select from a variety of options.
  All available image formats suppported by host libraries are supported.

  Draw-Order functions have been added to the Page menu as a fly-out using
  Scheme, allowing users to manipulate the order in which objects are drawn.

  The pop-up context menu now provides options for Copy, Move, Rotate, Mirror,
  Component Properties, Zoom to Mag, and Zoom to Selection. Additional options
  have also been add as fly-out side menus to the pop-up menu. Fly-out options
  use the X-Y location of the pointer when the context was invoked. Copy, Move,
  Rotate, and Mirror actions were reinstated and work identically to previous
  implementations.

  Additional icons are provided for "Move", Simulations, DRC, and attribute-
  reset. The default icon for "Move" was changed to have a more consistence
  theme with other editing icons.

  Menu tool-tips for File/Recent and Sessions/Restore now respond properly to
  user preferences. Tool-tips for symbol names on the Component dialog field
  shows the full path to symbol file.

  Anti-alias options were added to the Preference dialog in Gschem for the
  anti-alias method to be used by the chosen render. Modifing the anti-alias
  method longer requires a re-start for the new setting to take effect.

* Page Select Dialog has been enhanced to allow displaying file names without
  the paths, a toggle switch allows users to switch the path portion on or off
  in the tree-view. The user perference is retained between sessions.

* The appearance of Auto-Number Dialog has been improved and is more robust.
  The dialog was enhanced with frames and separate filter options for scope
  text including options for wild-cards and a list of recently used filters.
  Recent filters are not retained between sessions.

* The Line Type Dialog now includes an options combo to set the line end-type.


utils/scripts/pads_backannotate.pl
#!/usr/bin/perl -w
#
# Copyright (C) 2003, 2011-2012 Dan McMahill
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if  not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA

# This script is used to process PADS PowerPCB .eco files to backannotate
# changes to gEDA schematics.  Not all ECO file sections are implemented,
# and this program is relatively untested.  Please proceed with caution.
#

# for parsing input options
use Getopt::Long;

# for ceil function
use POSIX;

# don't allow -he to be interpreted as --help
$Getopt::Long::autoabbrev=0;

# make the options be case sensitive
$Getopt::Long::ignorecase=0;

&GetOptions( ('h|help' => \&usage,
	      'n|nocopy' => \$nocopy,
	      'v|verbose' => \$verbose,
	      'V|version' => \&version
	     ) );

usage() if $Getopt::Long::error;
usage() unless @ARGV;

# ECO file name
$eco = shift( @ARGV );

# if no schematic names follow, exit
usage() unless @ARGV;

# make sure the input netlist exists and we can open it
$i = 0;
while( @ARGV ) {
  $fname[$i] = shift( @ARGV );
  die "Schematic file $fname[$i] does not exist or can not be read"
    unless -r $fname[$i];
  $i++;
}

$filecnt = $i;


if( $verbose ){ print "Loading ECO file: $eco\n"; }
$eco_state="DEFAULT";
$got_end=0;
open( ECO, "$eco" ) or die "Can't open ECO file $eco !\n";

$line = <ECO>;
if( $line =~ /^\*PADS-ECO-V3\.0\*[\n\r]*$/ ) {
    print "Read ECO header line\n" if $verbose ;
} elsif( $line =~ /^\*PADS-ECO-\*/ ) {
    print "Unknown ECO file version: $line\n";
    exit 1;
} else {
    print "Invalid ECO file header: $line\n";
    exit 1;
}

while( $line = <ECO> ) {
    #  if( $verbose ){ print "$line\n"; }
    if( $line =~ /^\*CHGPART\*/ ) {
	# change type of part (forward annotation)
	$eco_state="CHGPART";
	next;
    } elsif( $line =~ /^\*DEL_ATTRIBUTE\*/ ) {
	# delete object attribute (forward annotation)
	$eco_state="DEL_ATTRIBUTE";
	next;
    } elsif( $line =~ /^\*DELPART\*/ ) {
	# delete part from the design (forward annotation)
	$eco_state="DELPART";
	next;
    } elsif( $line =~ /^\*DELPIN\*/ ) {
	# delete pins from net (forward annotation)
	$eco_state="DELPIN";
	next;
    } elsif( $line =~ /^\*END\*/ ) {
	# end of ECO file
	$eco_state="END";
	$got_end=1;
	next;
    } elsif( $line =~ /^\*JOINNET\*/ ) {
	# join two nets together (forward annotation)
	$eco_state="JOINNET";
	next;
    } elsif( $line =~ /^\*NET\*/ ) {
	# add pin to net (forward annotation)
	$eco_state="NET";
	next;
    } elsif( $line =~ /^\*PART\*/ ) {
	# add part to the design (forward annotation)
	$eco_state="PART";
	next;
    } elsif( $line =~ /^\*REMARK\*/ ) {
	# don't change state on comment lines
	print "Skipping *REMARK* line\n" if $verbose;
	next;
    } elsif( $line =~ /^\*RENNET\*/ ) {
	# rename netname (forward annotation)
	$eco_state="RENNET";
	next;
    } elsif( $line =~ /^\*RENPART\*/ ) {
	# rename part in design (forward/backward annotation)
	$eco_state="RENPART";
	print "Starting *RENPART* (Refdes Renumber) section\n" if $verbose;
	next;
    } elsif( $line =~ /^\*SET_ATTRIBUTE\*/ ) {
	# set object attribute (forward annotation)
	$eco_state="SET_ATTRIBUTE";
	next;
    } elsif( $line =~ /^\*SIGNAL\*/ ) {
	# part of *NET* and also *SPLITNET* (forward annotation)
	$eco_state="SIGNAL";
	next;
    } elsif( $line =~ /^\*SPLITNET\*/ ) {
	# split net into two new nets (forward annotation)
	$eco_state="SPLITNET";
	next;
    } elsif( $line =~ /^\*SWPGATES\*/ ) {
	$eco_state="SWPGATES";
	print "Starting *SWPGATES* (Gate Swapping) section\n" if $verbose;
	next;
    } elsif( $line =~ /^\*SWPPINS\*/ ) {
	$eco_state="SWPPINS";
	print "Starting *SWPPINS* (Pin Swapping) section\n" if $verbose;
	next;
    } elsif( $line =~ /^\*/ ) {
	print "WARNING:  Unknown command line:\n";
	print "          $line\n";
	$eco_state="DEFAULT";
	next;
    } else {
	# this must be a data line
	#if( $verbose ){ print "Processing data line: $line"; }
    }

    if( $eco_state =~ "RENPART" ) {
	parseRENPART($line);
    } elsif( $eco_state =~ "SWPGATES" ) {
	parseSWPGATES($line);
    } elsif( $eco_state =~ "SWPPINS" ) {
	parseSWPPINS($line);
    } else {
	print "Skipping lines in $eco_state section\n" if $verbose;
    }

}
close( ECO );

for($i=0; $i < $filecnt; $i++) {
  print "Processing schematic file #", $i+1, ": $fname[$i]\n";
  open(NETLIST,"$fname[$i]") or
    die "Can't open schematic $fname[$i]: $!\n";

  # open output netlist
  $outfname="$fname[$i].new";
  open(OUTSCH,">$outfname") or die "Can't open $outfname: $!\n";

    while($line = <NETLIST>) {
	$line = executeRENPART($line);
	print OUTSCH "$line";
    }
    close(NETLIST);
    close(OUTSCH);

    if( $nocopy ) {
      print "Leaving page #",$i+1," output in $outfname\n";
    }
    else {
      system("mv $outfname $fname[$i]");
    }

}

print "\n---- Gate Swapping ----\n";
executeSWPGATES();
print "\n---- Pin  Swapping ----\n";
executeSWPPINS();
print "\nBackannotation finished ", scalar localtime, "\n";

exit;


#######################################################################
#
# Subroutines
#
#######################################################################

#---------------------------------
# executeRENPART(line)
#---------------------------------

sub executeRENPART {
  my $line = shift(@_);

  return $line unless defined %cmd_rename;
  return $line unless $line =~ /^refdes=/;

  # pick out the reference designator
  $refdes = $line;
  $refdes =~ s/^refdes=//;
  $refdes =~ s/[\r\n]*$//;

  # see if its listed in our hash of reference designators to be
  # renamed
  return $line unless exists $cmd_rename{$refdes};

  print "executeRENPART():  Renaming $refdes to $cmd_rename{$refdes}\n"
      if $verbose;
  return "refdes=$cmd_rename{$refdes}\n";

}

#---------------------------------
# executeSWPGATES()
#---------------------------------

sub executeSWPGATES {
  my $key;

  foreach $key (keys %cmd_swap_gates ) {
    print "Please manually swap gates $key and  $cmd_swap_gates{$key}\n";
  }
}

#---------------------------------
# executeSWPPINS()
#---------------------------------

sub executeSWPPINS {
  my $key;
  my @pins;

  foreach  $key (keys %cmd_swap_pins ) {
    @pins = split '\.',,$cmd_swap_pins{$key};
    print "Please manually swap pins $pins[0] and $pins[1] on $key\n";
  }
}

#---------------------------------
# parseRENPART(line)
#---------------------------------

sub parseRENPART {
  my $line = shift(@_);
  my @refs;
  @refs = split ' ',,$line;

  print "parseRENPART():  Scheduling rename of $refs[0] to $refs[1]\n"
      if $verbose;
  $cmd_rename{$refs[0]} = $refs[1];

}

#---------------------------------
# parseSWPGATES(line)
#---------------------------------

sub parseSWPGATES {
  my $line = shift(@_);
  my @refs;
  @refs = split ' ',,$line;

  print "parseSWPGATES():  Scheduling swap of gate $refs[0] with $refs[1]\n"
      if $verbose;
  $cmd_swap_gates{$refs[0]} = $refs[1];

}

#---------------------------------
# parseSWPPINS(line)
#---------------------------------

sub parseSWPPINS {
  my $line = shift(@_);
  @refs = split ' ',,$line;
  @pins = split '\.',,$refs[1];

  print "parseSWPPINS():  Scheduling swap of pins on $refs[0] : pins ",
  "$pins[0] and $pins[1]\n" if $verbose;
  $cmd_swap_pins{$refs[0]} = $refs[1];

}

#---------------------------------
# usage()
#
# prints program usage
#---------------------------------

sub usage {
  my $pname = $0;
  $pname =~ s/.*\///g;

  print "Usage:\n\n";
  print "\t$pname [-n|--nocopy] [-v|--verbose] change.eco file1.sch [file2.sch [file3.sch ... ] ]\n";
  print "\t$pname -h|--help\n";
  print "\t$pname -V|--version\n";
  print "\n";
  print "$pname reads a PADS PowerPCB ECO file and backannotates changes\n";
  print "to a gEDA schematic.\n";
  print "\n";
  print "$pname accepts the following options:\n";
  print "\n";
  print "    -h|--help      Displays this help message.\n";
  print "\n";
  print "    -n|--nocopy    If given, this flag leaves the modified files in new files\n";
  print "                whose names are generated by appending a \".new\" to the\n";
  print "                original file names.  The default is to overwrite the original.\n";
  print "\n";
  print "    -v|--verbose   Enables verbose output.\n";
  print "\n";
  print "    -V|--version   Shows the version of this program.\n";
  print "\n\n";
  print "$pname was written by Dan McMahill <dmcmahill\@netbsd.org>\n";
  print "\n\n";
  exit;
}

#---------------------------------
# version()
#
# prints program version
#---------------------------------

sub version {
  open(PROG,"$0") or die "Could not open \"$0\" to find version\n\n";
  my $pname = $0;
  $pname =~ s/.*\///g;

  while($line = <PROG>) {
    if( $line =~ /^#\s*\$Id.*\$/) {
      @words = split ' ',,$line;
      $version = $words[3];
      $date = $words[4];
      print "$pname ($0):  Version $version, $date\n";
      exit;
    }
  }
  print "Could not determine version of \"$0\"\n\n";
  exit;
}

# ----------------------------------------------
#
# Change Log
#
# $Log$
# Revision 1.1  2003-08-31 21:39:52  ahvezda
# Added pads_backannotate perl script
#
# Revision 1.4  2003/07/19 04:34:55  dan
# bring this program up to a usable state.  Refdes renumbering is fully
# supported and pin/gate swapping while not supported, gives a report of
# the manual swaps needed.
#
# Revision 1.3  2003/03/03 03:44:26  dan
# Finish up code to handle the RENPART .eco directive.  The output files
# are still not written, but everything else is in place to handle backannotation
# of reference designator renumbers.
#
# Revision 1.2  2003/02/21 12:30:30  dan
# fix perl path
#
# Revision 1.1  2003/02/19 23:31:50  dan
# start (non functional so far) of a pads to gschem backannotation tool
#
#
#
